utils/metrics/metrics.go

```diff
@@ -0,0 +1,187 @@
+	"strings"
+)
+
+// TODO Add test cases for this file
+
+type Method int
+type Status int
+type Kind int
+type Protocol int
+
+const (
+	METHOD_GET Method = iota
+	METHOD_HEAD
+	METHOD_PUT
+	METHOD_OTHER
+)
+
+const (
+	OK Status = iota
+	NOT_FOUND
+	OTHER_STATUS
+)
+
+const (
+	AC Kind = iota
+	CAS
+)
+
+const (
+	HTTP Protocol = iota
+	GRPC
+)
+
+func (e Method) String() string {
+	// Actually HTTP names, but can be conceptually mapped also to GRPC protocol.
+	if e == METHOD_GET {
+		return "GET"
+	}
+	if e == METHOD_HEAD {
+		return "HEAD"
+	}
+	if e == METHOD_PUT {
+		return "PUT"
+	}
+	return "other"
+}
+
+func (e Status) String() string {
+	// Names that works for both HTTP and GRPC, instead of HTTP or GRPC specific codes.
+	if e == OK {
+		return "OK"
+	}
+	if e == NOT_FOUND {
+		return "NotFound"
+	}
+	return "other"
+}
+
+func (e Kind) String() string {
+	if e == AC {
+		return "AC"
+	}
+	if e == CAS {
+		return "CAS"
+	}
+	return "other"
+}
+
+type Metrics interface {
+	// TODO Document interface
+	IncomingRequestCompleted(kind Kind, method Method, status Status, headers map[string][]string, protocol Protocol)
+}
+
+type metrics struct {
+	categoryValues               map[string]map[string]struct{}
+	counterIncomingCompletedReqs *prometheus.CounterVec
+}
+
+func NewMetrics(config *config.Metrics) Metrics {
+
+	labels := []string{"method", "status", "kind"}
+	categoryValues := make(map[string]map[string]struct{})
+
+	if config != nil && config.Categories != nil {
+		for categoryName, whiteListedValues := range config.Categories {
+			// Normalize to lower case since canonical for gRPC headers
+			// and convention for prometheus.
+			categoryName := strings.ToLower(categoryName)
+
+			// Store white listed category values as set for efficient access
+			whiteListedSet := make(map[string]struct{})
+			for _, categoryValue := range whiteListedValues {
+				whiteListedSet[categoryValue] = struct{}{}
+			}
+			categoryValues[categoryName] = whiteListedSet
+
+			// Construct a prometheus label for each category.
+			// Prometheus does not allow changing set of
+			// labels until next time bazel-remote is
+			// restarted.
+			labels = append(labels, categoryName)
+		}
+	}
+
+	// For now we only count AC requests, and only the most common status codes,
+	// becuse:
+	//
+	//  - No identified use case for others.
+	//  - Limit number of prometheus time series (if many configured categories).
+	//  - Reduce performance overhead of counters (if many configured categories).
+	//  - Would otherwise require injecting invocations in more places.
+	//
+	// But the naming, and the labels, of the counter, are generic to allow
+	// counting additional requests types or status codes in the future. Without
+	// having to rename the counter and get issues with non continous history of
+	// metrics.
+
+	counterIncomingCompletedReqs := promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "bazel_remote_incoming_requests_completed_total",
```

Comment:
I guess this should have "ac" or "action_cache" or similar in the name?

Reply:
The logic in metrics.go is generic and there is a field called ’kind’ for the prometheus counter. Since http.go only forwards AC, that field will in practice always get the value AC.

My idea was to keep it generic, if someone will se a need to use that counter also for CAS, or something else, in the future. And without then renaming the counter (which would make it harder to see historical trends in prometheus database).

I use an expression like the following, to count all AC requests of method GET or HEAD:
 
`bazel_remote_incoming_requests_completed_total{kind="AC",method=~"GET|HEAD"}`

But I can remove the kind field and instead add AC to the counter name, if you think that is better?

Reply:
This sounds ok, I didn't spot the "kind".